samples/matter/common/src/diagnostic/diagnostic_logs_provider.cpp

```diff
@@ -0,0 +1,427 @@
+	case IntentEnum::kNetworkDiag:
+#ifdef CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_TEST
+		logSize = GetTestingLogsSize(intent);
+#endif /* CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_TEST */
+		/* TODO: add support for Network Diag Intent */
+		break;
+#ifdef CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_CRASH_LOGS
+	case IntentEnum::kCrashLogs:
+		logSize = GetCrashLogsSize();
+		break;
+#endif /* CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_CRASH_LOGS */
+	default:
+		break;
+	}
+
+	ChipLogDetail(Zcl, "Current Log size for Intent: %s is %zu", GetIntentStr(intent), logSize);
+
+	return logSize;
+}
+
+CHIP_ERROR DiagnosticLogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan &outBuffer,
+						  Optional<uint64_t> &outTimeStamp,
+						  Optional<uint64_t> &outTimeSinceBoot)
+{
+	LogSessionHandle sessionHandle = kInvalidLogSessionHandle;
+
+	CHIP_ERROR err = StartLogCollection(intent, sessionHandle, outTimeStamp, outTimeSinceBoot);
+	VerifyOrExit(CHIP_NO_ERROR == err, );
+
+	bool unusedOutIsEndOfLog;
+	err = CollectLog(sessionHandle, outBuffer, unusedOutIsEndOfLog);
+	VerifyOrExit(CHIP_NO_ERROR == err, );
+
+	err = EndLogCollection(sessionHandle);
+
+exit:
+	if (CHIP_NO_ERROR != err) {
+		outBuffer.reduce_size(0);
+	}
+
+	return err;
+}
+
+CHIP_ERROR DiagnosticLogProvider::Init()
+{
+#ifdef CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_SAVE_CRASH_TO_SETTINGS
+	ReturnErrorOnFailure(MoveCrashLogsToNVS());
+#endif
+
+	return CHIP_NO_ERROR;
+}
+
+void emberAfDiagnosticLogsClusterInitCallback(chip::EndpointId endpoint)
+{
+	auto &logProvider = DiagnosticLogProvider::GetInstance();
+
+	DiagnosticLogsServer::Instance().SetDiagnosticLogsProviderDelegate(endpoint, &logProvider);
+}
+
+#ifdef CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_CRASH_LOGS
+
+CHIP_ERROR DiagnosticLogProvider::LoadCrashData(CrashData *crashData)
+{
+	VerifyOrReturnError(crashData, CHIP_ERROR_INVALID_ARGUMENT);
+
+#ifdef CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_SAVE_CRASH_TO_SETTINGS
+	VerifyOrReturnError(Nrf::GetPersistentStorage().NonSecureHasEntry(&mCrashLogsStorageNode),
+			    CHIP_ERROR_NOT_FOUND);
+	size_t outSize = 0;
+	VerifyOrReturnError(Nrf::GetPersistentStorage().NonSecureLoad(
+				    &mCrashLogsStorageNode, reinterpret_cast<void *>(crashData->GetDescription()),
+				    sizeof(*crashData->GetDescription()), outSize),
+			    CHIP_ERROR_INTERNAL);
+	VerifyOrReturnError(outSize == sizeof(*crashData->GetDescription()), CHIP_ERROR_INTERNAL);
+#else
+	int ret = retention_is_valid(kCrashDataMem);
+	VerifyOrReturnError(ret == 1, CHIP_ERROR_NOT_FOUND);
+	ret = retention_read(kCrashDataMem, 0, reinterpret_cast<uint8_t *>(crashData->GetDescription()),
+			     sizeof(*crashData->GetDescription()));
+	VerifyOrReturnError(0 == ret, System::MapErrorZephyr(ret));
+#endif /* CONFIG_NCS_SAMPLE_MATTER_DIAGNOSTIC_LOGS_SAVE_CRASH_TO_SETTINGS */
+
+	return CHIP_NO_ERROR;
+}
+
+size_t DiagnosticLogProvider::GetCrashLogsSize()
+{
+	CrashData *crashData = Platform::New<CrashData>();
```

Comment:
Can't we just use a local object for that?

Reply:
This is going to be refactored anyway (I have it locally with regards to the next task). In the new version the single object will be used statically.